pkg/auth/server/selectprovider/selectprovider_test.go

```diff
@@ -0,0 +1,221 @@
+		resp := httptest.NewRecorder()
+		provider, handled, err := selectProvider.SelectAuthentication(testCase.Providers, resp, &http.Request{})
+
+		if err != nil {
+			t.Errorf("%s: unexpected error: %#v", k, err)
+			continue
+		}
+
+		if testCase.ExpectHandled != handled {
+			t.Errorf("%s: unexpected value for 'handled': %#v", k, handled)
+			continue
+		}
+
+		if testCase.ExpectSelectedProviderID != provider.ID {
+			t.Errorf("%s: unexpected provider ID: %#v", k, provider.ID)
+			continue
+		}
+
+		if len(testCase.ExpectContains) > 0 {
+			data, _ := ioutil.ReadAll(resp.Body)
+			body := string(data)
+			for i := range testCase.ExpectContains {
+				if !strings.Contains(body, testCase.ExpectContains[i]) {
+					t.Errorf("%s: did not find expected value %s: %s", k, testCase.ExpectContains[i], body)
+					continue
+				}
+			}
+		}
+	}
+}
+
+func TestValidateSelectProviderTemplate(t *testing.T) {
+	testCases := map[string]struct {
+		Template      string
+		TemplateValid bool
+	}{
+		"default provider selection template": {
+			Template:      defaultSelectProviderTemplateString,
+			TemplateValid: true,
+		},
+		"provider selection template example": {
+			Template:      SelectProviderTemplateExample,
+			TemplateValid: true,
+		},
+		"original provider selection template example": {
+			Template:      originalSelectProviderTemplateExample,
+			TemplateValid: true,
+		},
+		"template only prints first provider URL": {
+			Template:      invalidSelectProviderTemplate,
+			TemplateValid: false,
+		},
+	}
+
+	for k, testCase := range testCases {
+		allErrs := ValidateSelectProviderTemplate([]byte(testCase.Template))
+		if testCase.TemplateValid {
+			for _, err := range allErrs {
+				t.Errorf("%s: template validation failed when it should have succeeded: %v", k, err)
+			}
+		} else if len(allErrs) == 0 {
+			t.Errorf("%s: template validation succeeded when it should have failed", k)
+		}
+	}
+}
+
+// Make sure the original version of the default template always validates
+// this is to avoid breaking existing customized templates.
+const originalSelectProviderTemplateExample = `<!DOCTYPE html>
+<!--
+
+This template can be modified and used to customize the provider selection page. To replace
+the provider selection page, set master configuration option oauthConfig.templates.providerSelection to
+the path of the template file. Don't remove parameters in curly braces below.
+
+oauthConfig:
+  templates:
+    providerSelection: templates/select-provider-template.html
+
+The ID is unique for each provider and can be used for provider specific customizations like
+the example below.
+-->
+<html>
+  <head>
+    <title>Login</title>
+    <style type="text/css">
+      body {
+        font-family: "Open Sans", Helvetica, Arial, sans-serif;
+        font-size: 14px;
+        margin: 15px;
+      }
+    </style>
+  </head>
+  <body>
+
+    {{ range $provider := .Providers }}
+      <div>
+        {{ if eq $provider.ID "anypassword" }}
```

Comment:
add a comment that this is just an example of doing something custom for a particular provider